src/bin/stratum-miner.rs

```diff
@@ -0,0 +1,343 @@
+// Copyright 2018 The Grin Developers
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Stratum Miner, Connects to a grin stratum server, Gets new blocks and
+//! mines them, Returns solutions
+
+extern crate grin_core as core;
+extern crate grin_pow as pow;
+extern crate grin_util as util;
+extern crate rand;
+
+use std::mem;
+use std::thread;
+use std::time::Duration;
+use std::net::{Shutdown, TcpStream};
+use std::io::{ErrorKind, Read, Write};
+use std::sync::mpsc::Sender;
+use std::sync::mpsc::Receiver;
+use std::sync::mpsc::TryRecvError;
+use std::sync::mpsc::channel;
+use rand::Rng;
+
+use pow::cuckoo;
+use pow::MiningWorker;
+use core::ser;
+use core::core::{Block, BlockHeader};
+use core::consensus;
+
+// Receive a block from the parent, and mine it.  Check for a new block
+// occasionally.  Send solutions back to the parent.  Exit on failure,
+// the parent will re-spawn.
+fn run_mining_thread(
+	id: usize,
+	miner: &mut cuckoo::Miner,
+	s: Sender<BlockHeader>,
+	r: Receiver<Block>,
+	cuckoo_size: u32,
+) {
+	let debug_output_id = String::from("Miner1");
+	let mut b: Block;
+
+	println!("(Server ID: {}) Waiting for a block....", id);
+
+	// Blocking wait for an initial block
+	loop {
+		let blk = r.recv();
+		if blk.is_err() {
+			// I can't recover from a bad comm channel with the parent thread
+			return;
+		} else {
+			println!("Thread {} got a block: {:?}", id, blk);
+		}
+		b = blk.unwrap();
+		// dont mine a zero block.
+		if b.header.height != 0 {
+			break;
+		}
+	}
+
+	println!( "(Server ID: {}) Mining at Cuckoo{} at difficulty {} at height {} with nonce: {}: run_mining_thread",
+    debug_output_id,
+    cuckoo_size,
+    b.header.difficulty,
+    b.header.height,
+    b.header.nonce
+  );
+
+	loop {
+		// Do mining for a bit
+		for _ in 1..42 {
```

Comment:
This should all be using cuckoo-miner in async mode, which will just run on another thread and report when solutions are found, and you can send cancel messages to clients when a block has been found or time has expired.